### PR TITLE
Feature/basic setup book of habits api

### PR DESCRIPTION
**Pull request by AATerekhov**
#17 issues 
Added a state model according to the latest scheme

### CODE REVIEW

File: BookOfHabits/BookOfHabits.sln [hunk 1]
```diff
@@ -3,7 +3,25 @@ Microsoft Visual Studio Solution File, Format Version 12.00
 # Visual Studio Version 17
```
Comment (YauhenIlyushenka) [inline BookOfHabits/BookOfHabits.sln:3]: First of all, before creating PR you should always check target branch for merging. You need to switch from 'main' to 'develop' branch, because with accordance to Git flow, we work with 'develop' branch only on step of development.

Additionally, you need to check all files for merging in order to not allow appearing inappropriate files which are not directly related to your changes.

And could you please mention number of task in the name of PR? For instance 'Task 22 Add EF for Diary microservice'

File: BookOfHabits/BookOfHabitsMicroservice.Domain.Entity/Enums/OptionTemplate.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace BookOfHabitsMicroservice.Domain.Entity.Enums
+{
+    [Flags]
+    public enum OptionTemplate
+    {
```
Comment (YauhenIlyushenka) [inline BookOfHabits/BookOfHabitsMicroservice.Domain.Entity/Enums/OptionTemplate.cs:11]: Use a plural type name for an enumeration with bit fields - https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-classes-structs-and-interfaces#naming-enumerations

File: BookOfHabits/BookOfHabitsMicroservice.Domain.Entity/Enums/OptionHabit.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace BookOfHabitsMicroservice.Domain.Entity.Enums
+{
+    [Flags]
+    public enum OptionHabit
+    {
```
Comment (YauhenIlyushenka) [inline BookOfHabits/BookOfHabitsMicroservice.Domain.Entity/Enums/OptionHabit.cs:11]: Use a plural type name for an enumeration with bit fields - https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-classes-structs-and-interfaces#naming-enumerations

File: BookOfHabits/BookOfHabitsMicroservice.Domain.Entity/Enums/WeekDay.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace BookOfHabitsMicroservice.Domain.Entity.Enums
+{
+    [Flags]
+    public enum WeekDay
```
Comment (YauhenIlyushenka) [inline BookOfHabits/BookOfHabitsMicroservice.Domain.Entity/Enums/WeekDay.cs:10]: Use a plural type name for an enumeration with bit fields - https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-classes-structs-and-interfaces#naming-enumerations

File: BookOfHabits/BookOfHabitsMicroservice.Domain.ValueObjects/RoomName.cs [hunk 5]
```diff
@@ -0,0 +1,21 @@
+﻿namespace BookOfHabitsMicroservice.Domain.ValueObjects
+{
+    public class RoomName
+    {
+        public string Name { get; }
+        public RoomName(string name)
+        {
```
Comment (YauhenIlyushenka) [inline BookOfHabits/BookOfHabitsMicroservice.Domain.ValueObjects/RoomName.cs:7]: As a rule, the validation of fields should be executed in validators in order to not allow processing invalid data in API.

File: BookOfHabits/BookOfHabitsMicroservice.Domain.ValueObjects/Base/KeyName.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace BookOfHabitsMicroservice.Domain.ValueObjects.Base
+{
+    public abstract class KeyName : IEquatable<KeyName>, IEquatable<string>
+    {
+        public string Name { get; }
+        protected KeyName(string name)
+        {
```
Comment (YauhenIlyushenka) [inline BookOfHabits/BookOfHabitsMicroservice.Domain.ValueObjects/Base/KeyName.cs:13]: As a rule, the validation of fields should be executed in validators in order to not allow processing invalid data in API.

File: Diary/Diary.sln [hunk 7]
```diff
@@ -3,7 +3,11 @@ Microsoft Visual Studio Solution File, Format Version 12.00
 # Visual Studio Version 17
```
Comment (YauhenIlyushenka) [inline Diary/Diary.sln:3]: Remove it because it is not directly related to Ainur's microservice

File: Diary/Domain.Entities/BaseTypes/BaseEntity.cs [hunk 8]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
```
Comment (YauhenIlyushenka) [inline Diary/Domain.Entities/BaseTypes/BaseEntity.cs:1]: Remove it because it is not directly related to Ainur's microservice

File: Diary/Domain.Entities/BaseTypes/EventTypes.cs [hunk 9]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
```
Comment (YauhenIlyushenka) [inline Diary/Domain.Entities/BaseTypes/EventTypes.cs:1]: Remove it because it is not directly related to Ainur's microservice

File: Diary/Domain.Entities/BaseTypes/RelatedEntityTypes.cs [hunk 10]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
```
Comment (YauhenIlyushenka) [inline Diary/Domain.Entities/BaseTypes/RelatedEntityTypes.cs:1]: Remove it because it is not directly related to Ainur's microservice

File: Diary/Domain.Entities/Domain.Entities.csproj [hunk 11]
```diff
@@ -0,0 +1,9 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Comment (YauhenIlyushenka) [inline Diary/Domain.Entities/Domain.Entities.csproj:1]: the same

File: Diary/Domain.Entities/User.cs [hunk 12]
```diff
@@ -0,0 +1,12 @@
+﻿using Domain.Entities.BaseTypes;
+
```
Comment (YauhenIlyushenka) [inline Diary/Domain.Entities/User.cs:2]: the same. On the whole, it will be resolved by switching from 'main' to 'develop' as target branch for PR.

Check all your files for merging and get rid of files which is not related with your microservice.

File: BookOfHabits/BookOfHabitsMicroservice.Domain.Entity/Enums/OptionCoins.cs [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+namespace BookOfHabitsMicroservice.Domain.Entity.Enums
+{
+    [Flags]
+    public enum OptionCoins
+    {
+        None = 0,
+        FallingLevel = 1,
+        AutoPayment = 2,
+        CheckingLimit = 4,
+        CheckingReport = 8
```
Comment (Ainur555) [inline BookOfHabits/BookOfHabitsMicroservice.Domain.Entity/Enums/OptionCoins.cs:16]: Why do you need specific numbers for enums? Why don't you want to use the default?

File: BookOfHabits/BookOfHabitsMicroservice.Domain.Repository.Abstractions/IRepository.cs [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+namespace BookOfHabitsMicroservice.Domain.Repository.Abstractions
+{
+    public interface IRepository<TEntity, in TId> where TEntity : Entity<TId> where TId : struct
+    {
+        Task<IEnumerable<TEntity>> GetAllAsync();
+        Task<TEntity?> GetByIdAsync(TId id);
+        Task<TEntity> AddAsync(TEntity entity);
+        Task UpdateAsync(TEntity entity);
+        Task DeleteAsync(TEntity entity);
+        Task DeleteAsync(TId id);
```
Comment (Ainur555) [inline BookOfHabits/BookOfHabitsMicroservice.Domain.Repository.Abstractions/IRepository.cs:12]: As Danila told me, all asynchronous methods must have a cancellation token

Comment (AATerekhov) [reply BookOfHabits/BookOfHabitsMicroservice.Domain.Repository.Abstractions/IRepository.cs:12]: I agree too
